Fixed #30400 -- Improved typography of user facing strings.

jdufresne opened the pull request:
https://code.djangoproject.com/ticket/30400

jdufresne commented:
I'm not quite sure how to fix the test failure:

```

======================================================================
FAIL [0.025s]: test_help_text_translation (auth_tests.test_forms.SetPasswordFormTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-bionic/database/postgres/label/bionic-pr/python/python3.7/django/test/utils.py", line 370, in inner
    return func(*args, **kwargs)
  File "/home/jenkins/workspace/pull-requests-bionic/database/postgres/label/bionic-pr/python/python3.7/tests/auth_tests/test_forms.py", line 573, in test_help_text_translation
    self.assertIn(french_text, html)
AssertionError: 'Votre mot de passe ne peut pas trop ressembler à vos autres informations personnelles.' not found in '<p><label for="id_new_password1">Nouveau mot de passe:</label> <input type="password" name="new_password1" required id="id_new_password1"> <span class="helptext"><ul><li>Your password can’t be too similar to your other personal information.</li><li>Votre mot de passe doit contenir au minimum 12 caractères.</li></ul></span></p>\n<p><label for="id_new_password2">Confirmation du nouveau mot de passe:</label> <input type="password" name="new_password2" required id="id_new_password2"></p>'
```

I believe I'll need to update the translations, but I'm not the correct way to do that for the purposes of this test.

claudep commented:
You can try to pull https://github.com/django/django/commit/9500b69112c6

jdufresne commented:
That worked. Thanks!

jdufresne commented:
I was thinking that perhaps a note of some kind could be added to: https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#miscellaneous

But I was struggling to come up with useful phrasing. Thoughts?

jdufresne commented on django/core/validators.py:
```diff
@@ -466,8 +466,8 @@ def __eq__(self, other):
-        "File extension '%(extension)s' is not allowed. "
-        "Allowed extensions are: '%(allowed_extensions)s'."
+        "File extension “%(extension)s” is not allowed. "
+        "Allowed extensions are: “%(allowed_extensions)s”."
```
I'm not sure quotes around the list of allowed extensions makes. Thoughts on dropping them?

felixxm replied:
Agree.

jdufresne commented:
@claudep I hit the same issue with test:

```
======================================================================
FAIL [0.050s]: test_logentry_change_message_formsets (admin_utils.test_logentry.LogEntryTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-bionic/database/postgres/label/bionic-pr/python/python3.7/tests/admin_utils/test_logentry.py", line 141, in test_logentry_change_message_formsets
    "Modification de domain. Ajout de article « Added article ». "
AssertionError: 'Modification de domain. Added article “A[93 chars]le”.' != "Modification de domain. Ajout de article[144 chars]a0»."
- Modification de domain. Added article “Added article”. Changed title for article “Changed Title”. Deleted article “Title second article”.
+ Modification de domain. Ajout de article « Added article ». Modification de title pour l'objet article « Changed Title ». Suppression de article « Title second article ».
```

Could you update that translation for me as well? Feel free to force push directly to my branch if you'd like. Thanks.

Other than that, I believe this change is complete.

jdufresne commented on django/contrib/admin/models.py:
```diff
@@ -73,14 +73,14 @@ def __repr__(self):
             }
         elif self.is_deletion():
-            return gettext('Deleted "%(object)s."') % {'object': self.object_repr}
+            return gettext('Deleted “%(object)s.”') % {'object': self.object_repr}
```
Should this period be moved outside the quotes?

claudep replied:
Looks like in English the period is inside the quotes (see grammar sites).

jdufresne replied:
This is inconsistent with the "Added" case a few lines above as well as several other lines in the diff. Should we normalize these sentences to period inside quotes?

claudep replied:
Yes, consistency matters :-) Maybe @timgraham can bring his expertise here.

timgraham replied:
In American English, the punctuation usually goes inside the quote.

jdufresne replied:
Does that mean we should normalize to this style as part of this change?

blueyed replied:
Seems like it.

But it feels strange to me to have `Deleted "Bob."` instead of `Deleted "Bob".` - or am I missing something here?

jdufresne replied:
I agree that it _feels_ strange. I'll go with whatever is understood to be the most correct grammar style for American English.

jdufresne commented on django/contrib/admin/models.py:
```diff
@@ -73,14 +73,14 @@ def __repr__(self):
+            return gettext('Added “%(object)s”.') % {'object': self.object_repr}
         elif self.is_change():
-            return gettext('Changed "%(object)s" - %(changes)s') % {
+            return gettext('Changed “%(object)s” - %(changes)s') % {
```
Is this intended to be an em dash? I think so.

claudep replied:
Sure.

felixxm commented:
buildbot

felixxm commented on django/contrib/admin/options.py:
```diff
@@ -1501,7 +1501,7 @@ def _get_obj_does_not_exist_redirect(self, request, opts, object_id):
         and return a redirect to the admin index page.
         """
-        msg = _("""%(name)s with ID "%(key)s" doesn't exist. Perhaps it was deleted?""") % {
+        msg = _("""%(name)s with ID “%(key)s” doesn’t exist. Perhaps it was deleted?""") % {
```
`"""` -> `'`

felixxm commented on django/contrib/auth/password_validation.py:
```diff
@@ -202,4 +202,4 @@ def validate(self, password, user=None):
 
     def get_help_text(self):
-        return _("Your password can't be entirely numeric.")
+        return _("Your password can’t be entirely numeric.")
```
In all such cases we can change to single quotes.

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -2170,7 +2170,7 @@ def test_history_view_bad_url(self):
         self.assertEqual(
             [m.message for m in response.context['messages']],
-            ["""article with ID "foo" doesn't exist. Perhaps it was deleted?"""]
+            ["""article with ID “foo” doesn’t exist. Perhaps it was deleted?"""]
```
`"""` -> `'`

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -3689,7 +3689,7 @@ def test_change_view(self):
                 self.assertEqual(
                     [m.message for m in response.context['messages']],
-                    ["""empty model with ID "1" doesn't exist. Perhaps it was deleted?"""]
+                    ["""empty model with ID “1” doesn’t exist. Perhaps it was deleted?"""]
```
`"""` -> `'` etc.

jdufresne commented:
Thanks for the review @felixxm

Changes:

- Removed quotes from FileExtensionValidator’s allowed extensions list.
- Updated all newly identified cases. Thanks for finding these.
- Changed Python quote delimiter where appropriate across the diff. Including both Python files and template files.

> I believe that changes in French translations will disappear after rebase

As there were no merge conflicts and the new changes don't affect tests, this was not an issue.

felixxm commented:
@jdufresne Thanks!

@claudep Are you fine with updating French translations :fr: as a part of this patch? :crossed_fingers:

claudep commented:
Why would you want to update the translations in this patch? Any testing issue?

felixxm commented:
I'm talking only about these few lines that are already included in it.

claudep commented:
Ah sorry, I thought you were asking me to complete some translations. In fact, they could be removed from this patch if tests don't depend on them. Otherwise, no worries, they don't hurt.

felixxm commented:
@claudep Thanks, I limited changes only to necessary ones.

Aiky30 commented:
This change makes django 2.2 and 3.x compatible string checks difficult because the characters are different. 

We had: 

```
        self.assertEqual(
            message,
            'poll content with ID "144" doesn\'t exist. Perhaps it was deleted?',
        )
```

Now the message contains different characters for 2.2 and 3.x so fixing one version breaks compatibility for the other. 

The only way that I can think to patch this in third parties is to string replace the expected / test string for the different versions. :-(

jdufresne commented:
> The only way that I can think to patch this in third parties is to string replace the expected / test string for the different versions.

This is exactly what I have done and it wasn't bad. The check can be dropped overtime as your project removes EOL Django support.

claudep commented:
...or use `assertRegex` instead.
